File: 1905-count-sub-islands/1905-count-sub-islands.js
/**
 * @param {number[][]} grid1
 * @param {number[][]} grid2
 * @return {number}
 */
var countSubIslands = function(grid1, grid2) {
    const rows = grid1.length
    const cols = grid1[0].length
    let res = 0
    for (let r=0; r < rows; r++) {
        for (let c=0; c < cols; c++) {
            if (grid2[r][c] === 1) {
                res += dfs(r, c)
            }
        }
    }
    return res
    
    function dfs(r, c) {
        if (!(r in grid2) || !(c in grid2[r]) || grid2[r][c] !== 1) return 1
        let mask = 1
        grid2[r][c] = 0
        mask &= dfs(r+1, c)
        mask &= dfs(r-1, c)
        mask &= dfs(r, c+1)
        mask &= dfs(r, c-1)
        return mask & grid1[r][c]
    }
};

File: 1905-count-sub-islands/1905-count-sub-islands.py
class Solution:
    def countSubIslands(self, grid1: List[List[int]], grid2: List[List[int]]) -> int:
        
        rows = len(grid1)
        cols = len(grid1[0])
        
        def dfs(r: int, c: int) -> int:
            
            if r < 0 or r == rows or c < 0 or c == cols or grid2[r][c] == 0:
                return 1
            
            grid2[r][c] = 0
            mask = 1
            
            mask = mask & dfs(r+1, c)
            mask = mask & dfs(r-1, c)
            mask = mask & dfs(r, c-1)
            mask = mask & dfs(r, c+1)
            
            return mask & grid1[r][c]
        
        res = 0
        
        for r in range(rows):
            for c in range(cols):
                if grid2[r][c] == 1:
                    res += dfs(r, c)
                    
        return res
            